### DISCUSSION TITLE
Raw colour sample output not working

### DISCUSSION THREAD

**Discussion by njdoyle** [ImageMagick/ImageMagick#6725] [Contribute]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am using the latest version of ImageMagick
- [x] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
I'm trying to output raw pixel data for individual colour channels, it either isn't working or I'm misunderstanding how to use the feature.

I've reduced my use case down to this example. I want to take an input image (in this case the built-in logo, but any colour input will do), get the raw channel data for red, green, and blue, and then view that data as greyscale. (In my more involved case, I actually use the raw data for other things).

I expect the images `img.red.bmp`, `img.green.bmp`, and `img.blue.bmp` to be grayscale versions of each of the three colour channels of the input image but instead I get three identical images.

### Steps to Reproduce
```
magick logo: -depth 8 R:img.red
magick logo: -depth 8 G:img.green
magick logo: -depth 8 B:img.blue

magick -size 640x480 -depth 8 GRAY:img.red   img.red.bmp
magick -size 640x480 -depth 8 GRAY:img.green img.green.bmp
magick -size 640x480 -depth 8 GRAY:img.blue  img.blue.bmp
```

### System Configuration

- ImageMagick version: 7.0.10-55
- Environment (Operating system, version and so on): macOS 10.15.7 (19H114)
- Additional information:  
```
% magick -list configure

Path: [built-in]

Name                  Value
-------------------------------------------------------------------------------
DELEGATES             bzlib freetype heic jng jp2 jpeg lcms ltdl png tiff webp xml zlib
FEATURES              Cipher DPC HDRI
MAGICK_TEMPORARY_PATH /var/folders/16/mwv1bclx7vd3qx7qt9gs4ps80000gn/T
NAME                  ImageMagick
QuantumDepth          Q16
```

**Comment by fmw42** [comment]
I do not think your syntax of prefix with channels is permitted. Add -separate to your command to process your logo: and you will get 3 grayscale outputs

In place of

```
magick logo: -depth 8 R:img.red
magick logo: -depth 8 G:img.green
magick logo: -depth 8 B:img.blue

```

do the following


```
magick logo: -depth 8  -channel r -separate img.red
magick logo: -depth 8  -channel g -separate img.green
magick logo: -depth 8  -channel b -separate img.blue
```

**Comment by njdoyle** [comment]
These formats are listed here, so if they aren't permitted there is at least a documentation bug:
```
% magick -list format
   Format  Module    Mode  Description
———————————————————————————————————————
…
        B* RAW       rw+   Raw blue samples
…
        G* RAW       rw+   Raw green samples
…
        R* RAW       rw+   Raw red samples
```

Your alternative example outputs a GIF image and not raw sample data. What would be the correct format to use to get the raw sample data?

**Comment by fmw42** [comment]
I am on Mac OSX Sierra with ImageMagick 7.0.10-55 Q16 and doing 

magick -list format

does not show those options. It only shows

```
     BGR* BGR       rw+   Raw blue, green, and red samples
     BGRA* BGR       rw+   Raw blue, green, red, and alpha samples
     BGRO* BGR       rw+   Raw blue, green, red, and opacity samples

``` 
which process all three channels together and outputs a 3 channel raw image.


I have never seen the use of B:raw_image, with the preface of R: raw_image or B: raw_image or G: raw_image and not believe that one can separate channels that way for any type of image.

Are you on Windows or Linus or Mac? Also what is our version of ImageMagick?

Fred

**Comment by njdoyle** [comment]
```
 % magick -version
Version: ImageMagick 7.0.10-55 Q16 x86_64 2021-01-05 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI
Delegates (built-in): bzlib freetype heic jng jp2 jpeg lcms ltdl png tiff webp xml zlib
```

Downloaded and built myself earlier today before opening the bug.

I'm on macOS Catelina but can reproduce on Linux as well.

**Comment by njdoyle** [comment]
For reference:
```
% magick -list format | grep RAW
        A* RAW       rw+   Raw alpha samples
        B* RAW       rw+   Raw blue samples
        C* RAW       rw+   Raw cyan samples
    DCRAW  DNG       r--   Raw Photo Decoder (dcraw)
      ERF  DNG       r--   Epson RAW Format
        G* RAW       rw+   Raw green samples
        K* RAW       rw+   Raw black samples
        M* RAW       rw+   Raw magenta samples
        O* RAW       rw+   Raw opacity samples
        R* RAW       rw+   Raw red samples
      RAF  DNG       r--   Fuji CCD-RAW Graphic File
      RAW  DNG       r--   Raw
      X3F  DNG       r--   Sigma Camera RAW Picture File
        Y* RAW       rw+   Raw yellow samples
```

**Comment by njdoyle** [comment]
See:
* https://github.com/ImageMagick/ImageMagick/blob/master/coders/raw.h
* https://github.com/ImageMagick/ImageMagick/blob/master/coders/raw.c

**Comment by fmw42** [comment]
This is what I have for -version

```
Version: ImageMagick 7.0.10-55 Q16 x86_64 2021-01-04 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1) 
Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype gslib gvc jbig jng jp2 jpeg lcms lqr ltdl lzma openexr png ps raw rsvg tiff webp x xml zlib

```


This is what I have for

`magick -list format | grep -i raw`


```
      ARW  DNG       r--   Sony Alpha Raw Image Format
      BGR* BGR       rw+   Raw blue, green, and red samples
     BGRA* BGR       rw+   Raw blue, green, red, and alpha samples
     BGRO* BGR       rw+   Raw blue, green, red, and opacity samples
     CMYK* CMYK      rw+   Raw cyan, magenta, yellow, and black samples
    CMYKA* CMYK      rw+   Raw cyan, magenta, yellow, black, and alpha samples
      CR2  DNG       r--   Canon Digital Camera Raw Image Format
      CR3  DNG       r--   Canon Digital Camera Raw Image Format
      CRW  DNG       r--   Canon Digital Camera Raw Image Format
      DCR  DNG       r--   Kodak Digital Camera Raw Image File
    DCRAW  DNG       r--   Raw Photo Decoder (dcraw)
      DDS* DDS       rw+   Microsoft DirectDraw Surface
     DXT1* DDS       rw+   Microsoft DirectDraw Surface
     DXT5* DDS       rw+   Microsoft DirectDraw Surface
      ERF  DNG       r--   Epson RAW Format
     GRAY* GRAY      rw+   Raw gray samples
    GRAYA* GRAY      rw+   Raw gray and alpha samples
   GROUP4* TIFF      rw-   Raw CCITT Group4
      IIQ  DNG       r--   Phase One Raw Image Format
      K25  DNG       r--   Kodak Digital Camera Raw Image Format
      KDC  DNG       r--   Kodak Digital Camera Raw Image Format
      M4V  VIDEO     rw+   Raw VIDEO-4 Video
      MEF  DNG       r--   Mamiya Raw Image File
     MONO* MONO      rw-   Raw bi-level bitmap
      MRW  DNG       r--   Sony (Minolta) Raw Image File
      NEF  DNG       r--   Nikon Digital SLR Camera Raw Image File
      NRW  DNG       r--   Nikon Digital SLR Camera Raw Image File
      ORF  DNG       r--   Olympus Digital Camera Raw Image File
      PCT* PICT      rw-   Apple Macintosh QuickDraw/PICT
     PICT* PICT      rw-   Apple Macintosh QuickDraw/PICT
      RAF  DNG       r--   Fuji CCD-RAW Graphic File
      RAW  DNG       r--   Raw
      RGB* RGB       rw+   Raw red, green, and blue samples
   RGB565* RGB       r--   Raw red, green, blue samples in 565 format
     RGBA* RGB       rw+   Raw red, green, blue, and alpha samples
     RGBO* RGB       rw+   Raw red, green, blue, and opacity samples
      RMF  DNG       r--   Raw Media Format
      RW2  DNG       r--   Panasonic Lumix Raw Image
      SR2  DNG       r--   Sony Raw Format 2
      SRF  DNG       r--   Sony Raw Format
      X3F  DNG       r--   Sigma Camera RAW Picture File
    YCbCr* YCbCr     rw+   Raw Y, Cb, and Cr samples
   YCbCrA* YCbCr     rw+   Raw Y, Cb, Cr, and alpha samples
```

I am not sure if the difference is due to my delegates installations include libraw, which I believe shows as "raw" in my list of delegates from magick -version listed at the top of this page

**Comment by njdoyle** [comment]
I specifically do not have the libraw delegate. Looks like there may be a (second separate) bug with a conflict between libraw and the internal unrelated raw functionality?

**Comment by fmw42** [comment]
Well, I am just puzzled why -list format shows such differences especially with regard to raw formats

**Comment by njdoyle** [comment]
Just to be clear so we're on the same page. The "raw" functionality I am talking about and trying to use is getting raw sample data and libraw is unrelated functionality for reading raw camera sensor data. They are independent things. If one affects the other, that would be a bug too, but a separate bug from my original post.

**Comment by snibgo** [comment]
> ... get the raw channel data for red, green, and blue, and then view that data as greyscale ...

What does "raw channel data" mean?

As a general answer, I would use "-separate", for example like this:
```
magick logo: -channel RGB -separate +channel out.bmp
```
This should make outputs out-0.bmp, out-1.bmp and out-2.bmp.

The three outputs can be combined to get the original image:
```
magick out-0.bmp out-1.bmp out-2.bmp -combine x.png
```

I haven't come across the prefix "R:" etc, and don't know what it does or should do.

**Comment by njdoyle** [comment]
> What does "raw channel data" mean?

What I want is output that is nothing but raw pixel data. Assuming `-depth 8` and an output type of `R:`, the first byte of the output would be the first pixel's red value scaled from 0-255, the second byte would be the second pixel's red value, etc…

Just the raw channel data; no header, no trailer, no interleaved other data.

This is what these format's actually used to do I'm pretty certain but I'm assuming that because they're rarely used, a regression slipped in (maybe quite a while ago) without notice.

The other examples aren't what I want because the goal is to not need to parse/decode PNGs, BMPs, GIFs, etc.

**Comment by snibgo** [comment]
I suggest:
```
magick in.png -channel RGB -separate +channel -depth 8 GRAY:x-%d.bin
```
This makes x-0.bin, x-1.bin and x-2.bin. Each has no header, one byte per pixel.

They can be combined, to re-make the original, with:
```
magick -size 267x233 -depth 8 GRAY:x-0.bin GRAY:x-1.bin GRAY:x-2.bin -combine x.png
```

**Comment by njdoyle** [comment]
@snibgo The issue with that approach is that, to my understanding, the values are potentially modified (for example from sRGB gamma to linear gray). This means I don't really have the raw data and, at depths like 8, we have also lost precision and introduced banding.

See the transformation here:
https://github.com/ImageMagick/ImageMagick/blob/master/coders/gray.c#L855

Additionally, even if the above did/does what I want, that still means the initial bug I posted exists and is exposed to end users.

**Comment by snibgo** [comment]
I assume your input is sRGB. The GRAYColorspace has the same transfer curve as sRGB, so there is no conversion to linear gray. We can check this:
```
magick xc:sRGB(10%,10%,10%) -set colorspace sRGB -colorspace Gray txt:
# ImageMagick pixel enumeration: 1,1,65535,gray
0,0: (6553.5)  #199A199A199A  gray(10%)
```
There is no conversion to linear.

If your input is not sRGB, you can "-set colorspace sRGB" immediately after reading it. IM will then treat the image as sRGB, so there will be no modification.

Compare that to LinearGray:
```
magick xc:sRGB(10%,10%,10%) -set colorspace sRGB -colorspace LinearGray txt:
# ImageMagick pixel enumeration: 1,1,65535,lineargray
0,0: (656.846)  #029102910291  lineargray(1.00228%)
```
There is a conversion to linear. As you say, recording linear values with only 8 bits is bad news.

Your comments about "R:" etc may point to a bug in IM.

**Comment by njdoyle** [comment]
You're right, it doesn't look like `GRAY:` linearizes anything. This means it is a possible workaround. Thank you.

I do think the original bug still stands though for the other formats in `raw.c`/`raw.h`.
